apis/apiextensions/v1beta1/composition_types.go

```diff
@@ -133,35 +212,145 @@ type ReadinessCheck struct {
 	MatchInteger int64 `json:"matchInteger,omitempty"`
 }
 
-// Patch is used to patch the field on the base resource at ToFieldPath
-// after piping the value that is at FromFieldPath of the target resource through
-// transformers.
+// patchSetMap stores a list of patches from a set, against the Set ID.
+type patchSetMap map[string]Patch
+
+// PatchType defines the usable inbuilt patch types
+type PatchType string
+
+// Predefined PatchTypes.
+const (
+	PatchTypeFromCompositeFieldPath PatchType = "FromCompositeFieldPath" // Default
+	PatchTypePatchSet               PatchType = "PatchSet"
+)
+
+// Patch objects are applied between the composite and composed resources.
+// Their exact behaviour depends on the Type selected.
+// If the given type does not match one of the standard patch types, then
+// it may refer to a customPatchSet defined on the Composition. In this case,
+// the referenced patch will be dereferenced into the patch list in its' place.
 type Patch struct {
+	// Type can be a standard PatchSet value, or a 'Name' which refers
+	// to a custom patch type defined on the composition.
+	// +optional
+	// +kubebuilder:validation:Enum=FromCompositeFieldPath;PatchSet
+	// +kubebuilder:default=FromCompositeFieldPath
+	Type PatchType `json:"type,omitempty"`
+
+	// PatchSet is a patch type that allows multiple nested patches
+	// +optional
+	*PatchSet `json:",inline"`
+
+	// PatchFromCompositeFieldPath is the 'standard' patch type that maps a field
+	// from the composite resource to the composed resource, with optional
+	// transformations.
+	// +optional
+	*PatchFromCompositeFieldPath `json:",inline"`
+}
 
+// Apply applies a patch between the from and to objects. Depending on the
+// patch type, this may be from the composite resource to the composed resource,
+// or vice-versa.
+func (c *Patch) Apply(from, to runtime.Object) error {
+	var patcher interface {
+		Apply(from, to runtime.Object) error
+	}
+	switch c.Type {
+	case PatchTypeFromCompositeFieldPath:
+		patcher = c.PatchFromCompositeFieldPath
+	case PatchTypePatchSet:
+		patcher = c.PatchSet
+	default:
+		return errors.Errorf(errInvalidPatchType, c.Type)
+	}
+
+	// An interface equals nil only if both the type and value are nil. Above,
+	// even if t.<Type> is nil, its type is assigned to "transformer" but we're
+	// interested in whether only the value is nil or not.
```

Comment:
Nit: Would it be simpler to check whether the `c.PatchSet` or `c.PatchFromCompositeFieldPath` structs are nil inside the `switch` to avoid needing to use reflection here?

Reply:
Rather than in the `switch`, what if it becomes the concern of the actual patch type `Apply` method? 

The patch types already need to check that their own required fields are set correctly, so I wonder if just moving the pointer nil check there makes sense - leaving the `Patch.Apply` as a dispatcher?

Reply:
That SGTM!